[libc++] Support using Newlib as libc with locale support.

wecing opened the pull request:
Tested on Newlib v4.5.0. The newlib.h was mostly copied from linux.h within the same directory.

Newlib locale support was initially added in D5385 [1], but removed in revision cfde4fb [2], because of revision 78b6374 [3].

[1]: https://reviews.llvm.org/D5385
[2]: https://github.com/llvm/llvm-project/pull/113721
[3]: https://reviews.llvm.org/D49927

wecing committed:
[libc++] Support using Newlib as libc with locale support.

Tested on Newlib v4.5.0. The newlib.h was mostly copied from linux.h
within the same directory.

Newlib locale support was initially added in D5385 [1], but removed in
revision cfde4fb [2], because of revision 78b6374 [3].

[1]: https://reviews.llvm.org/D5385
[2]: https://github.com/llvm/llvm-project/pull/113721
[3]: https://reviews.llvm.org/D49927

wecing committed:
clang-format fix

jroelofs reviewed (commented):
Looks fine to me, but you should get @ldionne’s blessing.

wecing committed:
Merge branch 'llvm:main' into newlib_locale_review

wecing committed:
Use cmake variables defined in commit a6643f2.

a6643f2 implemented a better version of the cmake variables we need;
this commit switches to using that, and removes the old version I
implemented.

philnik777 reviewed (commented):
Could you point out the changes compared to `linux.h`?

wecing committed:
Reduce divergence between newlib.h and linux.h

Many functions copied from linux.h are no longer needed after
https://github.com/llvm/llvm-project/commit/693f700e1dba5d601d3ac999ff5471a3d0aaac70.

Also removed some whitespace changes to make newlib.h closer to linux.h.

philnik777 commented:
If there is no actual difference I think it's fine to just update the `#if`. Maybe we can rename the header to something more appropriate.

ldionne commented on libcxx/include/__locale_dir/locale_base_api.h:
```diff
@@ -118,6 +118,8 @@
 #  elif defined(__linux__)
 #    include <__locale_dir/support/linux.h>
+#  elif _LIBCPP_LIBC_NEWLIB
+#    include <__locale_dir/support/newlib.h>
```
Is Newlib ever used on a platform where `__linux__` is defined? It seems to me like we might be checking the wrong thing by checking `elif defined(__linux__)` above, perhaps what we want is `elif defined(_GLIBC)`?

Also CC @reactive-firewall

wecing replied:
Thank you for the review and insights, @ldionne and @reactive-firewall !

> AFAIK, newlib does seem to provide the strto*_l functions unconditionally (albeit not sure where in their codebase 🤷)

For the record, they are defined in the non-`_l` `.c` files, e.g. [newlib/libc/stdlib/strtol.c](https://sourceware.org/git/?p=newlib-cygwin.git;a=blob;f=newlib/libc/stdlib/strtol.c;h=09d4333ed05f497ec00b7192204644349441e58d;hb=HEAD#l227) for `strtol_l`.

> perhaps what we want is `elif defined(_GLIBC)`?

Per my discussion with @philnik777 above, `linux.h` and `newlib.h` are essentially the same. So it sounds like we need something like this:

```
#  elif defined(_GLIBC) || _LIBCPP_LIBC_NEWLIB
#    include <__locale_dir/support/minimal_libc.h>
```

And rename `linux.h` to `minimal_libc.h`?

ldionne reviewed (approved):
Based on my other comment, LGTM. I'd rather refactor this myself with the bigger picture than break the consistency right now.

ldionne committed:
Update libcxx/include/__locale_dir/support/newlib.h
